overlord/fdestate/fdestate.go

```diff
@@ -66,3 +79,289 @@ func EFISecureBootDBUpdateCleanup(st *state.State) error {
+}
+
+// KeyslotRoleInfo stores information about a key slot role
+type KeyslotRoleInfo struct {
+	// PrimaryKeyID is the ID for the primary key found in
+	// PrimaryKeys field of FdeState
+	PrimaryKeyID int `json:"primary-key-id"`
+	// Parameters is indexed by container role name
+	Parameters map[string]KeyslotRoleParameters `json:"params,omitempty"`
+	// TPM2PCRPolicyRevocationCounter is the handle for the TPM
+	// policy revocation counter.  A value of 0 means it is not
+	// set.
+	TPM2PCRPolicyRevocationCounter uint32 `json:"tpm2-pcr-policy-revocation-counter,omitempty"`
+}
+
+type hashAlg crypto.Hash
+
+// UnmarshalJSON implements json.Unmarshaler.UnmarshalJSON
+func (h *hashAlg) UnmarshalJSON(b []byte) error {
+	var s string
+	if err := json.Unmarshal(b, &s); err != nil {
+		return err
+	}
+	switch s {
+	case "sha256":
+		*h = hashAlg(crypto.SHA256)
+	default:
+		return fmt.Errorf("unknown algorithm %s", s)
+	}
+
+	return nil
+}
+
+// MarshalJSON implements json.Marshaler.MarshalJSON
+func (h hashAlg) MarshalJSON() ([]byte, error) {
+	switch crypto.Hash(h) {
+	case crypto.SHA256:
+		return json.Marshal("sha256")
+	default:
+		return nil, fmt.Errorf("unknown algorithm %v", h)
+	}
+}
+
+// KeyDigest stores a HMAC(key, salt) of a key
+type KeyDigest struct {
+	// Algorithm is the algorithm for
+	Algorithm hashAlg `json:"alg"`
+	// Salt is the salt for the HMAC digest
+	Salt []byte `json:"salt"`
+	// Digest is the result of `HMAC(key, salt)`
+	Digest []byte `json:"digest"`
+}
+
+const defaultHashAlg = crypto.SHA256
+
+func getPrimaryKeyDigest(devicePath string) (KeyDigest, error) {
+	salt, digest, err := secbootGetPrimaryKeyHash(devicePath, crypto.Hash(defaultHashAlg))
+	if err != nil {
+		return KeyDigest{}, err
+	}
+
+	return KeyDigest{
+		Algorithm: hashAlg(defaultHashAlg),
+		Salt:      salt,
+		Digest:    digest,
+	}, nil
+}
+
+func (kd *KeyDigest) verifyPrimaryKeyDigest(devicePath string) (bool, error) {
+	return secbootVerifyPrimaryKeyHash(devicePath, crypto.Hash(kd.Algorithm), kd.Salt, kd.Digest)
+}
+
+// PrimaryKeyInfo provides information about a primary key that is used to manage key slots
+type PrimaryKeyInfo struct {
+	Digest KeyDigest `json:"digest"`
+}
+
+// FdeState is the root persistent FDE state
+type FdeState struct {
+	// PrimaryKeys are the keys on the system. Key with ID 0 is
+	// reserved for snapd and is populated on first boot. Other
+	// IDs are for externally managed keys.
+	PrimaryKeys map[int]PrimaryKeyInfo `json:"primary-keys"`
+
+	// KeyslotRoles are all keyslot roles indexed by the role name
+	KeyslotRoles map[string]KeyslotRoleInfo `json:"keyslot-roles"`
+}
+
+const fdeStateKey = "fde"
+
+func initializeState(st *state.State) error {
+	var s FdeState
+	err := st.Get(fdeStateKey, &s)
+	if err == nil {
+		// TODO: Do we need to do something in recover?
+		return nil
+	}
+
+	if !errors.Is(err, state.ErrNoState) {
+		return err
+	}
+
+	dataUUID, dataErr := disksDMCryptUUIDFromMountPoint(dirs.GlobalRootDir)
```

Comment:
this seems to be a bug I'm hitting right now. `/` will always be from the boot base, this likely should use `dirs.SnapdStateDir(dirs.GlobalRootDir)`

Reply:
Right. I think I am missing a spread test to verify the state. Both for classic and core.

Reply:
You can cherry pick https://github.com/canonical/snapd/pull/14615/commits/97f6d238bb4080a70413c3b57e546d4455c9d589 from #14615